### PR TITLE
Feature/ntgl 394 updated link sharing

### PR DESCRIPTION
**Pull request by incalescence**
### Description
- Enhanced link sharing experience now works with copying and pasting url!
- Easily share links to events with friends 
- Displays a preview of the event card that the user can add to their timetable
- Will not let user add if event (with same details) already exists
- Modal is a nested route with URL in the format notangles.csesoc.app/event/{my-custom-event-encrypted}
### Demo
Notangles - Timetable Planner - Google Chrome 2023-06-29 12-35-37

### CODE REVIEW

File: client/src/components/timetable/ExpandedEventView.tsx [hunk 1]
```diff
@@ -193,6 +193,10 @@ const ExpandedEventView: React.FC<ExpandedEventViewProps> = ({
     delete updatedEventData[id];
     setCreatedEvents(updatedEventData);
   };
+
+  // link sharing url for custom events
+  const url = 'localhost:3000/event/' + btoa(JSON.stringify(eventPeriod));
```
Comment (manhualu) [inline client/src/components/timetable/ExpandedEventView.tsx]: will this work with the hardcoded 'localhost:3000' url when we deploy it?

Comment (manhualu) [reply client/src/components/timetable/ExpandedEventView.tsx]: I think you'll have to get the current URL of the webpage

Comment (manhualu) [reply client/src/components/timetable/ExpandedEventView.tsx]: should be able to do something like: 
```suggestion
const currentUrl = window.location.href;
const url = currentUrl + 'event/' + btoa(JSON.stringify(eventPeriod));
```

File: client/src/components/EventShareModal.tsx [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+      setErrorVisibility(true);
+      return;
+    }
+    const newEvent = createEventObj(name, location, description, color, day, startTime, endTime);
+    setCreatedEvents({
+      ...createdEvents,
+      [newEvent.event.id]: newEvent,
+    });
+
+    setEarliestStartTime(Math.min(Math.floor(earliestStartTime), Math.floor(startTime)));
```
Comment (manhualu) [inline client/src/components/EventShareModal.tsx:89]: these lines of code is also used in the CustomEvent.tsx file, can you put it into a function?

File: client/src/components/EventShareModal.tsx [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+  const [open, setOpen] = React.useState(true);
+  const [link, setLink] = useState('');
+  const [eventPreview, setEventPreview] = useState(false);
+  const [event, setEvent] = useState({ name: '', location: '', color: '' });
+  const { createdEvents, setCreatedEvents } = useContext(CourseContext);
+  const { setAlertMsg, setErrorVisibility, setDays, earliestStartTime, setEarliestStartTime, latestEndTime, setLatestEndTime } =
+    useContext(AppContext);
+
+  const updateDays = (day: number) => {
+    if (day == 5 || day == 6) {
```
Comment (manhualu) [inline client/src/components/EventShareModal.tsx]: put this in a shared file as well to use in CustomEvent.tsx

**Review by manhualu — DISMISSED** [review]

**Comment by manhualu** [comment]
@incalescence should we support multiple timetables for the adding of the event? As right now the user can only place an event on their first default timetable

File: client/src/components/controls/CustomEvent.tsx [hunk 4]
```diff
@@ -238,47 +213,13 @@ const CustomEvent: React.FC = () => {
...
-    setCreatedEvents({
-      ...createdEvents,
-      [newEvent.event.id]: newEvent,
-    });
-
-    setEarliestStartTime(Math.min(Math.floor(earliestStartTime), Math.floor(startTime)));
-    setLatestEndTime(Math.max(Math.ceil(latestEndTime), Math.ceil(endTime)));
-
-    updateDays(day);
+    if (daysShort.indexOf(day) == 5 || daysShort.indexOf(day) == 6) {
```
Comment (Rayahhhmed) [inline client/src/components/controls/CustomEvent.tsx]: Any reason why === wasnt used here?

File: client/src/components/controls/CustomEvent.tsx [hunk 5]
```diff
@@ -343,33 +283,22 @@ const CustomEvent: React.FC = () => {
...
         <ExecuteButton
           variant="contained"
           color="primary"
           disableElevation
           disabled={
-            (eventType === 'General' && (eventName === '' || eventDays.length === 0)) ||
-            (eventType === 'Tutoring' && (courseCode === '' || classCode === '')) ||
-            (eventType === 'Via Link' && link === '')
+            (eventType === 'General' && (eventName === '' || location === '' || eventDays.length === 0)) ||
+            (eventType === 'Tutoring' && (courseCode === '' || classCode === ''))
```
Comment (Rayahhhmed) [inline client/src/components/controls/CustomEvent.tsx]: All of these values are faulty code can be cleaned up using their boolean values ie
(eventType === 'General' && (!eventName || ...))

Comment (manhualu) [reply client/src/components/controls/CustomEvent.tsx]: damn this was past me 💀, will fix

Comment (manhualu) [reply client/src/components/controls/CustomEvent.tsx]: also just realised this does not implement location is optional 🤔 (merging dev messed up)

Comment (Rayahhhmed) [reply client/src/components/controls/CustomEvent.tsx]: issokay i forgive u
but next time i shout coffee (or tea) if u stuff up

**Comment by Rayahhhmed** [comment]
> @incalescence should we support multiple timetables for the adding of the event? As right now the user can only place an event on their first default timetable

How should this be, like the user then gets to decide which timetable this event gets placed on? 
Like they can select all or select couple ? that will be cool
Like a modal that pops up and they get to pick which timetable the event is at.

**Review by manhualu — APPROVED** [review]

**Review by Rayahhhmed — APPROVED** [review]

**Pull request merged**